File: src/main/java/pl/training/workshop/m5/s07_collapsehierarchy/step1/HallCatalog.java
package pl.training.workshop.m5.s07_collapsehierarchy.step1;

import java.util.List;

/** Krok 1: bez zmian. */
public final class HallCatalog {
    private final List<Hall> halls = List.of(
            new Hall("Sala 1", 12, 15, 10),
            new Hall("Sala 2", 10, 12, 9),
            new ImaxHall("Sala IMAX", 14, 22));

    public String describe(String name) {
        return halls.stream()
                .filter(hall -> hall.name().equals(name))
                .map(Hall::describe)
                .findFirst()
                .orElse("brak sali: " + name);
    }

    public boolean isVip(String name, int row) {
        return halls.stream().anyMatch(hall -> hall.name().equals(name) && hall.isVip(row));
    }
}

File: csharp/src/Training.Workshop/M5/S07CollapseHierarchy/Step3/HallCatalog.cs
namespace Training.Workshop.M5.S07CollapseHierarchy.Step3;

/// <summary>Krok 3: bez zmian - dlatego Safe Delete ImaxHall przeszło bez ostrzeżeń.</summary>
public sealed class HallCatalog
{
    private readonly IReadOnlyList<Hall> _halls =
    [
        new Hall("Sala 1", 12, 15, 10),
        new Hall("Sala 2", 10, 12, 9),
        Hall.Imax("Sala IMAX", 14, 22),
    ];

    public string Describe(string name)
    {
        return _halls
            .Where(hall => hall.Name == name)
            .Select(hall => hall.Describe())
            .FirstOrDefault() ?? "brak sali: " + name;
    }

    public bool IsVip(string name, int row)
    {
        return _halls.Any(hall => hall.Name == name && hall.IsVip(row));
    }
}
